tests/unit/test_util.py

```diff
@@ -10,8 +10,12 @@
 import unittest
 from unittest.mock import patch
 from io import StringIO
+from tempfile import TemporaryDirectory
 
+def test_function(arg):
+    return arg + arg
 
+anytree = sys.modules['anytree']
```

Comment:
This line is repeated at line 36 for some reason which I don't follow. Do we necessarily require `sys.modules['anytree']` there, or can we import `anytree` with `have_optional_dependency` there?

Reply:
Yes i felt this is necessary as if we don't do it `test_is_constant_and_can_evaluate` fails with:
```
ERROR: test_is_constant_and_can_evaluate (test_util.TestUtil)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/util.py", line 353, in have_optional_dependency
    module = importlib.import_module(module_name)
  File "/usr/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1032, in _find_and_load
ModuleNotFoundError: import of anytree halted; None in sys.modules

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/PyBaMM/PyBaMM/tests/testcase.py", line 29, in wrapped
    return method(*args, **kwargs)
  File "/home/runner/work/PyBaMM/PyBaMM/tests/unit/test_util.py", line 34, in test_is_constant_and_can_evaluate
    symbol = pybamm.PrimaryBroadcast(0, "negative electrode")
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/expression_tree/broadcasts.py", line 76, in __init__
    child = pybamm.Scalar(child)
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/expression_tree/scalar.py", line 29, in __init__
    super().__init__(name)
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/expression_tree/symbol.py", line 231, in __init__
    for x in self.pre_order()
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/expression_tree/symbol.py", line 526, in pre_order
    anytree = have_optional_dependency("anytree")
  File "/home/runner/work/PyBaMM/PyBaMM/pybamm/util.py", line 369, in have_optional_dependency
    raise ModuleNotFoundError(f"Optional dependency {module_name} is not available. See https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html#optional-dependencies for more details.")
ModuleNotFoundError: Optional dependency anytree is not available. See https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html#optional-dependencies for more details.
``` 

> can we import anytree with have_optional_dependency there?

I tried doing this but it was failing with the same logs.

Reply:
Okay, it should be fine to keep this. We can look into neater solutions later